### ISSUE TITLE
What The Hell Is Going On With SourceWindow?

### ISSUE THREAD

**Issue by Garethp** [joonty/vdebug#267]
I'm trying to add a feature where if you do `:q` on the main buffer (the code) that it closes all the other buffers and closes the connection. Slightly problem is that I can't seem to hook in to it's close event with `autocmd BufferWinLeave`. If I do `autocmd BufferWinLeave \* echo 'Closed'`, it'll echo for doing a `:q` on every buffer **except** the SourceWindow buffer. I'm not all *that* good at vimscript yet, and I'm still kind of trying to figure out the `vim` module and how your `WIndow` classes work, so I'm not sure what's happening here. Can you shed some light on how the windows are created, what's different about `SourceWindow` and why I can't hook in to it's closing with `autocmd`?

**Comment by Garethp** [comment]
Okay, so I know (kind of) how to reproduce this simply.

But opening vim, typing `:tabnew`, followed by `:1buffer` and then `:new Window`, we get the same result, where Buffer 1 is open in both Tab 1 and Tab 2. Doing a `au BufWinLeave * echo 'test'` and trying to do `:q` on Buffer 1 *inside* Tab 2 doesn't tigger the `autcmd`. Now... how do I fix this?

**Comment by Garethp** [comment]
Closing this as I have a solution. Basically what's hapening is that I want to hook in to the Window of the Buffer closing, but there's no `WinClose` event to hook in to. So, instead, I've set up a `WinEnter` event which then causes the debugger to check if the `SourceWindow` is still open, and if it's not, close the debugger. As part of the close, it removes it's `au` hook so the hook doesn't hang around. Crude, needs some optimisation, but I'll work on it
